src/test/java/com/commercetools/sync/commons/utils/CtpQueryUtilsTest.java

```diff
@@ -104,6 +111,47 @@ void queryAll_WithCustomSort_ShouldNotSortByIdAsc() {
                 getSecondPageQuery(keyPredicateFunction, categories, QuerySort.of("id desc")));
     }
 
+    @Test
+    void queryAll_WithGraphQlRequest_ShouldFetchPagedResult() {
+        // preparation
+        final SphereClient sphereClient = mock(SphereClient.class);
+        final List<ResourceKeyId> resultPage = IntStream
+            .range(0, 500)
+            .mapToObj(i -> mock(ResourceKeyId.class))
+            .collect(Collectors.toList());
+        final ResourceKeyId lastCategory = resultPage.get(resultPage.size() - 1);
+        when(lastCategory.getId()).thenReturn(UUID.randomUUID().toString());
+
+        final Set<ResourceKeyId> mockPage = resultPage.stream().collect(Collectors.toSet());
+
+        final GraphQlResult pagedGraphQlResult = mock(GraphQlResult.class);
+        when(pagedGraphQlResult.getResults())
+            .thenReturn(mockPage)
+            .thenReturn(mockPage.stream().limit(10).collect(Collectors.toSet()));
+
+        when(sphereClient.execute(any())).thenReturn(completedFuture(pagedGraphQlResult));
+
+        final Set<String> keysToQuery = IntStream
+            .range(1, 510)
```

Comment:
Can you make this number bigger, e.g. around 2000? The idea is to make more than 2 queries graphql queries and then check if predicate is still correct. We had a bug before where the query keeps expanding with new requests (probably not in this PR, but still good to check).

Reply:
f8d487e